### ISSUE TITLE
Users can un-ban some terms pretty easily.

### ISSUE THREAD

**Issue by MikaalSky** [Khan/live-editor#309]
var escape = (function(){return this;})();
try{
    escape.BabyHint.bannedProperties = [];
}catch(e){}
Unbans terms like document and such. Also, escape grants access to eval and some other great stuff.

**Comment by Gigabyte-Giant** [comment]
This code doesn't *actually* unban any objects/functions. Sure, you can get around BabyHint, but not by setting `bannedProperties` to an empty array.

**Comment by JavascriptFTW** [comment]
It seems someone was able to get around BabyHint by putting `random(/\/*/);` before their code (https://www.khanacademy.org/computer-programming/bens-module-system/4728010161913856).
Putting a multi-line comment after this statement reactivates BabyHint indicating that BabyHint is considering the body of the regular expression to be the opening of a multi-line comment.

**Comment by Gigabyte-Giant** [comment]
What a clever way of disabling things. This is sorta one of those "why didn't I think of that" things, right?

Anyhow, as ever, a stricter check [here](https://github.com/Khan/live-editor/blob/master/js/output/pjs/babyhint.js#L259-L281), should be able to solve this issue...

**Comment by kevinb7** [comment]
We should also have an AST pass that throws if someone includes `return this` in their code.

**Comment by JavascriptFTW** [comment]
@kevinb7 It should probably be more subtle than that since just throwing an error whenever we find `return this` would break the perfectly innocent practice of method chaining.

**Comment by kevinb7** [comment]
@JavascriptFTW good point.  We'd have to also determine if the function was a constructor or not which is a little trickier.

**Comment by kevinb7** [comment]
We could insert a check if `this === window` before return `this`.

**Comment by kevinbarabash** [comment]
If `this === window` then we should probably return an object that looks like `window`.  That way people can get use to some of JavaScript's peculiarities.  The more we can make our sandbox look like plain old JavaScript, the better it will be for our users.  Once we have this object that looks like `window` then we should also be able to allow users to access it directly via `window`.

**Comment by Gigabyte-Giant** [comment]
So, @kevinbarabash, just to make sure I'm understanding you correctly, you're suggesting doing something like this?:

Before:
```javascript
var myFunction = (function() {
    return this;
})();
```

After:
```javascript
var myFunction = (function() {
    if (this === window) {
        return { /* custom window object */ };
    }
    return this;
})();
```
Of course, in my example, I didn't include any of the other transforms/injects that take place.

If that is indeed what you're suggesting, then an AST transform should still work...

**Comment by kevinbarabash** [comment]
I think that should work.  The custom window object would probably just be `__env__` with a few items hidden.

**Comment by Potato42** [comment]
AST transforms are practically useless at stopping users from accessing the Window object:

    // jshint evil:true
    println(eval("window"));
    println(Object.constructor("return this")());

Why are we even trying?  The HTML environment has complete access to `window`.  It's not even frozen like the PJS environment.  Why can't we just do something like that for PJS?  Is there some vulnerability that keeps us from making the PJS environment a simple wrapper over the HTML one?

**Comment by kevinbarabash** [comment]
The reasons for trying to sandbox (at least in my mind) are:
- preventing safety features such as the loop protector from being disabled
- preventing modifications to globals that might break the environment

I agree that it's a bit of an uphill battle seeing as there are lots of ways to get access to `window`.  I think knowing about `window` is important so I'd rather have a safe version that users could access.

As for the difference between the pjs and webpage environments, I think it's a combination of pjs being more popular and more mature.  Having greater consistency between the two environments would be nice.

**Comment by Potato42** [comment]
We have the same problem with disabling loop protector/other safety features in the HTML environment.  In fact, since loop protector has been rather finicky lately, I've seen many users disabling it ([example](https://www.khanacademy.org/computer-programming/infinite-slide/6611264526745600)).  It makes me wonder if we could somehow put things like that in a different scope, but that's maybe a different issue.  
Breaking the environment is a problem to some extent, but it seems like either essentially the first problem or a "why would you even do that" thing.  I guess I don't see a big deal with breaking the environment anyway apart from the result possibly being unexpected.

**Comment by pamelafox** [comment]
I'm on the road now so I can't write much, but I wanted to pipe in with an
additional reason that we tried to disable window: messy/confusing code.
When I first joined KA, I noticed the hot list was full of programs with
externals hacks and hard-to-explain code that did DOM/window manip. My hope
was that people could view most programs on the hot list and mostly
understand the code in them, and having less programs with externals hacks
would help this.

I won't claim that my reasoning is correct as thats a longer discussion,
but I wanted to let you all in on some of the initial thought process.
